searx/engines/command.py

```diff
@@ -0,0 +1,189 @@
+'''
+searx is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+searx is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with searx. If not, see < http://www.gnu.org/licenses/ >.
+'''
+
+
+from os.path import expanduser, isabs, realpath, commonprefix
+from re import MULTILINE, search as re_search
+from shlex import split as shlex_split
+from subprocess import Popen, PIPE
+from time import time
+from threading import Thread
+
+from searx import logger
+from searx.utils import str_split
+
+
+offline = True
+paging = True
+command = []
+delimiter = {}
+parse_regex = {}
+query_type = ''
+query_enum = []
+working_dir = realpath('.')
+result_separator = '\n'
+result_template = 'key-value.html'
+timeout = 4.0
+
+_command_logger = logger.getChild('command')
+_compiled_parse_regex = {}
+
+
+def init(engine_settings):
+    check_parsing_options(engine_settings)
+
+    if 'command' not in engine_settings:
+        raise ValueError('engine command : missing configuration key: command')
+
+    global command, working_dir, result_template, delimiter, parse_regex, timeout
+
+    command = engine_settings['command']
+
+    if 'working_dir' in engine_settings:
+        working_dir = engine_settings['working_dir']
+        if not isabs(engine_settings['working_dir']):
+            working_dir = realpath(working_dir)
+
+    if 'parse_regex' in engine_settings:
+        parse_regex = engine_settings['parse_regex']
+        for result_key, regex in parse_regex.items():
+            _compiled_parse_regex[result_key] = re.compile(regex, flags=MULTILINE)
+    if 'delimiter' in engine_settings:
+        delimiter = engine_settings['delimiter']
+
+
+def search(query, params):
+    cmd = _get_command_to_run(query)
+    if not cmd:
+        return []
+
+    results = []
+    process = Popen(cmd, stdout=PIPE, stderr=PIPE)
+    try:
+        reader_thread = Thread(target=_get_results_from_process, args=(results, process, params['pageno']))
+        try:
+            reader_thread.start()
+            return_code = process.wait(timeout=timeout)
+            if return_code != 0:
+                raise RuntimeError('non-zero return code when running command', cmd, return_code, process.stderr.read())
+        finally:
+            reader_thread.join(timeout=timeout)
+    finally:
+        if process.is_alive():
+            process.kill()
```

Comment:
I'm not sure if that should be addressed:
* SIGKILL may leave some forked processes : https://stackoverflow.com/questions/21869242/killing-parent-process-along-with-child-process-using-sigkill (if ```cmd``` fork).
* https://stackoverflow.com/questions/32222681/how-to-kill-a-process-group-using-python-subprocess 
   * SIGHUP then SIGKILL might be better ?
   * or process group ?